rust/src/commands/transaction.rs

```diff
@@ -0,0 +1,171 @@
+    },
+    error::Result as DataFusionResult,
+    physical_plan::{
+        coalesce_partitions::CoalescePartitionsExec, common::compute_record_batch_statistics,
+        empty::EmptyExec, Distribution, ExecutionPlan, Partitioning, SendableRecordBatchStream,
+        Statistics,
+    },
+};
+use lazy_static::lazy_static;
+use std::sync::Arc;
+
+lazy_static! {
+    /// Schema expected form plans wrapped by transaction
+    pub static ref OPERATION_SCHEMA: ArrowSchema =
+        ArrowSchema::new(vec![ArrowField::new("serialized", DataType::Utf8, true,)]);
+}
+
+pub(crate) fn serialize_actions(actions: Vec<Action>) -> DataFusionResult<RecordBatch> {
+    let serialized = StringArray::from(
+        actions
+            .iter()
+            .map(serde_json::to_string)
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(to_datafusion_err)?,
+    );
+    Ok(RecordBatch::try_new(
+        Arc::new(OPERATION_SCHEMA.clone()),
+        vec![Arc::new(serialized)],
+    )?)
+}
+
+/// Write command
+#[derive(Debug)]
+pub struct DeltaTransactionPlan {
+    table_uri: String,
+    input: Arc<dyn ExecutionPlan>,
+    operation: DeltaOperation,
+    app_metadata: Option<serde_json::Map<String, serde_json::Value>>,
+}
+
+impl DeltaTransactionPlan {
+    /// Wrap partitioned delta operations in a DeltaTransaction
+    pub fn new<T>(
+        table_uri: T,
+        input: Arc<dyn ExecutionPlan>,
+        operation: DeltaOperation,
+        app_metadata: Option<serde_json::Map<String, serde_json::Value>>,
+    ) -> Self
+    where
+        T: Into<String>,
+    {
+        Self {
+            table_uri: table_uri.into(),
+            input: Arc::new(CoalescePartitionsExec::new(input)),
+            operation,
+            app_metadata,
+        }
+    }
+
+    /// Arrow schema expected to be produced by wrapped operations
+    pub fn input_operation_schema() -> ArrowSchemaRef {
+        Arc::new(OPERATION_SCHEMA.clone())
+    }
+}
+
+#[async_trait]
+impl ExecutionPlan for DeltaTransactionPlan {
+    /// Return a reference to Any that can be used for downcasting
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> ArrowSchemaRef {
+        Arc::new(OPERATION_SCHEMA.clone())
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![self.input.clone()]
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        Partitioning::UnknownPartitioning(1)
+    }
+
+    fn required_child_distribution(&self) -> Distribution {
+        Distribution::UnspecifiedDistribution
+    }
+
+    fn with_new_children(
+        &self,
+        _children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> DataFusionResult<Arc<dyn ExecutionPlan>> {
+        todo!()
+    }
+
+    async fn execute(&self, _partition: usize) -> DataFusionResult<SendableRecordBatchStream> {
+        let mut table =
+            get_table_from_uri_without_update(self.table_uri.clone()).map_err(to_datafusion_err)?;
```

Comment:
When implementing this, I also often thought, that decoupling `DeltaTransaction` from the actual table instance would be beneficial. Haven't really looked at it more closely, but is this something we should try in this PR?

Reply:
Yeah, I think it's something we can handle as follow ups since the scope of this PR is already quite large. We should leave a comment and link to a tracking issue though :)